src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -133,7 +468,88 @@ instance (Arithmetic a, KnownNat n) => BoolType (ByteString n (ArithmeticCircuit
                 flipped <- forM bits $ \b -> newAssigned (\p -> one - p b)
                 horner flipped
 
-    l || r = binOp l r (\i j x -> x i + x j - x i * x j)
+    l || r = bitwiseOperation l r (\i j x -> x i + x j - x i * x j)
+
+    l && r = bitwiseOperation l r (\i j x -> x i * x j)
+
+    xor l r = bitwiseOperation l r (\i j x -> x i + x j - (x i * x j + x i * x j))
+
 
-    l && r = binOp l r (\i j x -> x i * x j)
+instance
+  ( KnownNat wordSize
+  , KnownNat n
+  , wordSize <= n
+  , 1 <= wordSize
+  , 1 <= n
+  , Mod n wordSize ~ 0
+  , Arithmetic a
+  ) => ToWords (ByteString n (ArithmeticCircuit a)) (ByteString wordSize (ArithmeticCircuit a)) where
+
+    toWords (ByteString x xs) = bitsToBS <$> chunksOf (Haskell.fromIntegral $ minNumberOfRegisters @a @wordSize) (circuits solve)
+      where
+        bitsToBS :: [ArithmeticCircuit a] -> ByteString wordSize (ArithmeticCircuit a)
+        bitsToBS []     = error "toWords.bitsToBS :: unreachable"
+        bitsToBS (r:rs) = ByteString r rs
+
+        solve :: forall i m. MonadBlueprint i a m => m [i]
+        solve = do
+            bits <- toBits @n @a x xs
+            let words = chunksOf (Haskell.fromIntegral $ getNatural @wordSize) bits
+            wordsBits <- mapM (fromBits @wordSize @a) words
+            pure $ Haskell.concat wordsBits
+
+instance
+  ( KnownNat m
+  , KnownNat n
+  , m <= n
```

Comment:
Do you need this constraint? Logically it is implied from `Mod n m ~ 0` (except if `n == 0`)

Reply:
Logically, yes. But Haskell is unable to infer `n <= n + a `or `Mod (a * n) n ~ 0`.

Reply:
Oh wait, your question is about a different thing. Yeah, probably it is safe to remove this